File: Website/InfoTerraTerra/InfoTerraTerra_Library/Tracking/TrackingQrOpenStatistics.cs
using InfoTerraTerra_Library.Extensions;

namespace InfoTerraTerra_Library.Tracking;

public class TrackingQrOpenStatistics
{
    public SingleVolantinoStatistics[] SingleVolantinoStatistics { get; }
    public KeyValuePair<string, string>[] GlobalCounters { get; }

    public TrackingQrOpenStatistics(
        IReadOnlyCollection<TrackingGroupedData>? trackingGroupedData)
    {
        GlobalCounters = new KeyValuePair<string, string>[]
        {
            new("Numero totale QR aperti", "Nessuno 😭")
        };

        SingleVolantinoStatistics = Array.Empty<SingleVolantinoStatistics>();
            
        var globalCounters = new List<KeyValuePair<string, string>>();
        if (trackingGroupedData == null || !trackingGroupedData.Any()) return;
        
        {
            globalCounters.Add(new KeyValuePair<string, string>(
                "Numero totale QR aperti", trackingGroupedData.Sum(d => d.HowMany).ToString()));
            var qrApertiPerCitta = trackingGroupedData
                .GroupBy(d => d.Citta)
                .ToArray();
            if (qrApertiPerCitta.Any())
            {
                globalCounters.AddRange(qrApertiPerCitta
                    .Where(c => c.Key != null)
                    .OrderByDescending(c => c.Count())
                    .Select(c => new KeyValuePair<string, string>(
                        $"A {c.Key!.CapitalizeFirstLetter()}",
                        c.Sum(d => d.HowMany).ToString())));
            }

            GlobalCounters = globalCounters.ToArray();
        }

        {
            var singleVolantinoStatistics = new List<SingleVolantinoStatistics>();
            var qrApertiPerVolantino = trackingGroupedData
                .GroupBy(d => d.IdVolantino)
                .Where(d => d.Any())
                .ToArray();
            if (qrApertiPerVolantino.Any())
            {
                singleVolantinoStatistics.AddRange(qrApertiPerVolantino
                    .OrderByDescending(c => c.Key)
                    .Select(c =>
                    {
                        var counters = new KeyValuePair<string, string>[]
                        {
                            new(
                                "Numero QR aperti",
                                c.Sum(d => d.HowMany).ToString())
                        };
                            
                        var numeroQrApertiPerCitta = c
                            .GroupBy(d => d.Citta)
                            .Where(d => d.Key != null)
                            .Select(d => new KeyValuePair<string, string>(
                                d.Key!.CapitalizeFirstLetter(),
                                d.Sum(x => x.HowMany).ToString()))
                            .OrderBy(x => x.Key)
                            .ToArray();

                        var numeroQrApertiPerLuogo = c
                            .GroupBy(d => d.Luogo)
                            .Where(d => d.Key != null)
                            .Select(d => new KeyValuePair<string, string>(
                                d.Key!.CapitalizeFirstLetter(),
                                d.Sum(x => x.HowMany).ToString()))
                            .OrderBy(x => x.Key)
                            .ToArray();
                            
                        var numeroQrApertiPerCittaVia = c
                            .Where(x => !string.IsNullOrWhiteSpace(x.Via) && !string.IsNullOrWhiteSpace(x.Citta))
                            .GroupBy(d => $"{d.Citta} - {d.Via}")
                            .Select(d => new KeyValuePair<string, string>(
                                d.Key.CapitalizeFirstLetter(),
                                d.Sum(x => x.HowMany).ToString()))
                            .OrderBy(x => x.Key)
                            .ToArray();
                            
                        return new SingleVolantinoStatistics
                        {
                            TitoloVolantino = c.First().Volantino.Title,
                            Counters = counters,
                            PerCitta = numeroQrApertiPerCitta,
                            PerVia = numeroQrApertiPerCittaVia,
                            PerLuogo = numeroQrApertiPerLuogo
                        };
                    }));
            }
                
            SingleVolantinoStatistics = singleVolantinoStatistics.ToArray();
        }
    }
}

File: dataLayer/tracking/TrackingQrOpenStatistics.ts
import {SingleVolantinoStatistics} from "./SingleVolantinoStatistics";
import StringHelper from "../../services/StringHelper";
import {TrackingGroupedData} from "./TrackingGroupedData";
import {KeyValuePair} from "../../model/KeyValuePair";

export class TrackingQrOpenStatistics {

    singleVolantinoStatistics!: SingleVolantinoStatistics[];
    globalCounters: { key: string; value: string }[];

    constructor(trackingGroupedData: TrackingGroupedData[] | null) {

        // Counter globali
        const globalCounters: KeyValuePair<string, string>[] = [];

        console.log("trackingGroupedData", trackingGroupedData);

        if (!trackingGroupedData || trackingGroupedData.length === 0) {
            this.globalCounters = globalCounters;
            return;
        }

        globalCounters.push({
            key: "Numero totale QR aperti",
            value: trackingGroupedData.reduce((acc, d) => acc + d.HowMany, 0).toString()
        });

        const qrApertiPerCitta = trackingGroupedData
            .reduce((acc, d) => {
                const existing = acc.find(c => c.key === d.Citta);
                if (existing) {
                    existing.value += d.HowMany;
                }
                else {
                    acc.push({key: d.Citta!, value: d.HowMany});
                }
                return acc;
            }, [] as KeyValuePair<string, number>[]);

        if (qrApertiPerCitta.length > 0) {
            const sortedQrApertiPerCitta = qrApertiPerCitta
                .filter(c => c.key !== null)
                .sort((a, b) => b.value - a.value);

            globalCounters.push(
                ...sortedQrApertiPerCitta.map(c => ({
                    key: `A ${StringHelper.capitalizeFirstLetter(c.key!)}`,
                    value: c.value.toString()
                }))
            );
        }

        this.globalCounters = globalCounters;
        // Fine counter globali


        // Counter per volantino
        const singleVolantinoStatistics: SingleVolantinoStatistics[] = [];

        const qrApertiPerVolantino = (trackingGroupedData || [])
            .reduce((acc, d) => {
                const existing = acc.find(c => c.idVolantino === d.IdVolantino);
                if (existing) {
                    existing.data.push(d);
                }
                else {
                    acc.push({idVolantino: d.IdVolantino, data: [d]});
                }
                return acc;
            }, [] as { idVolantino: number, data: TrackingGroupedData[] }[]);

        if (qrApertiPerVolantino.length > 0) {
            singleVolantinoStatistics.push(
                ...qrApertiPerVolantino.map(c => {
                    const counters: KeyValuePair<string, string>[] = [
                        {key: "Numero QR aperti", value: c.data.reduce((acc, x) => acc + x.HowMany, 0).toString()}
                    ];

                    const numeroQrApertiPerCitta = c.data
                        .reduce((acc, d) => {
                            const existing = acc.find(x => x.key === d.Citta);
                            if (existing) {
                                existing.value += d.HowMany;
                            }
                            else {
                                acc.push({key: d.Citta!, value: d.HowMany});
                            }
                            return acc;
                        }, [] as KeyValuePair<string, number>[])
                        .filter(d => d.key !== null)
                        .map(d => ({key: StringHelper.capitalizeFirstLetter(d.key!)!, value: d.value.toString()}))
                        .sort((a, b) => a.key.localeCompare(b.key));

                    const numeroQrApertiPerLuogo = c.data
                        .reduce((acc, d) => {
                            const existing = acc.find(x => x.key === d.Luogo);
                            if (existing) {
                                existing.value += d.HowMany;
                            }
                            else {
                                acc.push({key: d.Luogo!, value: d.HowMany});
                            }
                            return acc;
                        }, [] as KeyValuePair<string, number>[])
                        .filter(d => d.key !== null)
                        .map(d => ({key: StringHelper.capitalizeFirstLetter(d.key!)!, value: d.value.toString()}))
                        .sort((a, b) => a.key.localeCompare(b.key));

                    const numeroQrApertiPerCittaVia = c.data
                        .filter(x => !!(x.Via && x.Citta))
                        .reduce((acc, d) => {
                            const key = `${d.Citta} - ${d.Via}`;
                            const existing = acc.find(x => x.key === key);
                            if (existing) {
                                existing.value += d.HowMany;
                            }
                            else {
                                acc.push({key: key, value: d.HowMany});
                            }
                            return acc;
                        }, [] as KeyValuePair<string, number>[])
                        .map(d => ({key: StringHelper.capitalizeFirstLetter(d.key!)!, value: d.value.toString()}))
                        .sort((a, b) => a.key.localeCompare(b.key));

                    return {
                        titoloVolantino: c.data[0].volantino.title,
                        counters: counters,
                        perCitta: numeroQrApertiPerCitta,
                        perVia: numeroQrApertiPerCittaVia,
                        perLuogo: numeroQrApertiPerLuogo
                    };
                })
            );
        }

        this.singleVolantinoStatistics = singleVolantinoStatistics;
        // Fine counter per volantino
    }
}
